packages/katello/katello/hostname-change.rb

```diff
@@ -239,6 +240,103 @@ def setup_opt_parser
+      old_fqdn = @old_hostname
+      soa_admin_domain = "root.#{zone}"
+      begin
+        query_dns = Resolv::DNS.new(nameserver: [local_ip], search: [zone], ndots: 1)
+        a_record = query_dns.getresource(old_fqdn, Resolv::DNS::Resource::IN::A)
+        existing_ip = a_record.address.to_s if a_record
+        soa_record = query_dns.getresource(zone, Resolv::DNS::Resource::IN::SOA)
+        existing_serial = soa_record.serial if soa_record
+        new_serial = (existing_serial + 1).to_s if existing_serial
+      rescue Resolv::ResolvError => e
+        STDOUT.puts e
+        STDOUT.puts 'Parsing zone file as fallback'
+        run_cmd('rndc freeze')
+        zone_file = "/var/named/dynamic/db.#{zone}"
+        zone_file_data = File.read(zone_file).split("\n")
+        a_record = zone_file_data.find { |line| line.include?(' A ') }
+        existing_ip = a_record.split(' A ').last if a_record
+        serial_line = zone_file_data.find { |line| line.include?('Serial') }
+        existing_serial = serial_line.split('Serial').first.to_i if serial_line
+        new_serial = (existing_serial + 1).to_s if existing_serial
+      ensure
+        run_cmd('rndc thaw')
+      end
+      new_fqdn = @new_hostname
+      key_file = @scenario_answers['foreman_proxy']['keyfile']
+      reverse_zone = @scenario_answers['foreman_proxy']['dns_reverse']
+      if reverse_zone.is_a?(Array)
+        reverse_zone = reverse_zone.first
+      end
+
+      {
+        local_ip: local_ip,
+        zone: zone,
+        old_fqdn: old_fqdn,
+        soa_admin_domain: soa_admin_domain,
+        existing_ip: existing_ip,
+        new_serial: new_serial,
+        new_fqdn: new_fqdn,
+        key_file: key_file,
+        reverse_zone: reverse_zone
+      }
+    end
+
+    def update_dns_records(dns_data = {})
+      dns_data.each { |k, v| raise "Error gathering DNS data: couldn't find value for '#{k}'" unless v }
+
+      local_ip = dns_data[:local_ip]
+      zone = dns_data[:zone]
+      old_fqdn = dns_data[:old_fqdn]
+      soa_admin_domain = dns_data[:soa_admin_domain]
+      existing_ip = dns_data[:existing_ip]
+      new_serial = dns_data[:new_serial]
+      new_fqdn = dns_data[:new_fqdn]
+      key_file = dns_data[:key_file]
+      reverse_zone = dns_data[:reverse_zone]
+
+      STDOUT.puts 'updating DNS records:'
+      # Use nsupdate to update DNS records
+      # Update SOA record to new hostname; add new A and NS records; delete old A and NS records.
+      # Multi-line strings are not indented because Ruby 2.0 doesn't support <<~ heredocs
+
+      forward_dns_command = <<-HEREDOC
+echo -e "local #{local_ip}
+zone #{zone}
+update add #{zone} 10800 SOA #{new_fqdn} #{soa_admin_domain}. #{new_serial} 86400 3600 604800 3600
+update add #{zone}. 3600 IN NS #{new_fqdn}.
+update delete #{zone}. IN NS #{old_fqdn}
+update delete #{old_fqdn} A
+update add #{new_fqdn} 86400 A #{existing_ip}
+send
+" | nsupdate -l -k #{key_file}
+      HEREDOC
```

Comment:
Some of this comes down to coding style, so this is just a suggestion, but what about just referencing the hash and creating separate methods?

```ruby
def forward_dns_command(dns_data)
<<-HEREDOC
echo -e "local #{dns_data[:local_ip]}
zone #{dns_data[:zone]}
update add #{dns_data[:zone]} 10800 SOA #{dns_data[:new_fqdn]} #{dns_data[:soa_admin_domain]}. #{dns_data[:new_serial]} 86400 3600 604800 3600
update add #{dns_data[:zone]}. 3600 IN NS #{dns_data[:new_fqdn]}.
update delete #{dns_data[:zone]}. IN NS #{dns_data[:old_fqdn]}
update delete #{dns_data[:old_fqdn]} A
update add #{dns_data[:new_fqdn]} 86400 A #{dns_data[:existing_ip]}
send
" | nsupdate -l -k #{dns_data[:key_file]}
      HEREDOC
end
```

and could even dry up the nsupdate flags in case they need updating later.

```ruby
def nsupdate_command(dns_entries, keyfile)
    "echo -e \"#{dns_entries}\" | nsupdate -l -k #{keyfile}"
end

def forward_dns_command(dns_data)
  dns_entries = <<-HEREDOC
local #{dns_data[:local_ip]}
zone #{dns_data[:zone]}
update add #{dns_data[:zone]} 10800 SOA #{dns_data[:new_fqdn]} #{dns_data[:soa_admin_domain]}. #{dns_data[:new_serial]} 86400 3600 604800 3600
update add #{dns_data[:zone]}. 3600 IN NS #{dns_data[:new_fqdn]}.
update delete #{dns_data[:zone]}. IN NS #{dns_data[:old_fqdn]}
update delete #{dns_data[:old_fqdn]} A
update add #{dns_data[:new_fqdn]} 86400 A #{dns_data[:existing_ip]}
send
      HEREDOC
  nsupdate_command(dns_entries, dns_data[:keyfile])
end

# do the same for reverse_dns_command
 ```

Reply:
@johnpmitsch yeah I just thought it was really ugly to have `dns_data` everywhere.   I feel like adding that 7 characters dozens of times kind of cancels out the benefit of separating out the methods.  But I'll change it if you feel strongly about it.

Reply:
@jeremylenz It would be my preference for readability, but I don't  see any technical or best-practices reason to do one over the other and we don't enforce method limits in this repo, so the final choice is up to you!

Reply:
@jeremylenz either way, I would consider DRYing up the nsupdate flags. I seem to remember a bug in k-c-hostname where we only updated cli flags in once place and missed another

Reply:
@johnpmitsch Added the `nsupdate_command` method per your suggestion.  Also separated out `query_dns_server` and `parse_zone_file` methods.

Also, I realized I needed to handle the situation where `Serial` is different for the forward and reverse zones, so we're handling that now too.